Integrate Continuous Rewards

dbrehmer opened the pull request:
Creating this PR so we have a place to discuss and make notes about changes.

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
transfering tests from tests branch

simusud committed:
fixing merge issues

simusud committed:
tests of laons,rewards and reserve fund

simusud committed:
imported existing tbears tests

simusud committed:
resolving merge conflicts

simusud committed:
rewards and reserve score tests update

simusud committed:
update method removed

simusud committed:
added tests cases

simusud committed:
base file to change day to 30 seconds

simusud committed:
small changes

simusud committed:
small changes

simusud committed:
individual rewards distribution issue fixed

simusud committed:
rewards distribution issue fixed

simusud committed:
added method to change day to 30 sec

simusud committed:
small update

simusud committed:
dao fund score tests

simusud committed:
liquidation test update

simusud committed:
test dummy_oracle for liquidation test

simusud committed:
rebalancer updates

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
test cases update

simusud committed:
rebalancer tests updates

simusud committed:
rebalancer tests updates

simusud committed:
rebalancing tests update

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' into loans-integration-tests

simusud committed:
rebalancing tests update

simusud committed:
loans update

simusud committed:
merge conflicts fixes

simusud committed:
merge conflict fixes

simusud committed:
rebalancing updates

simusud committed:
rebalancing updates

simusud committed:
solving merge conflicts

simusud committed:
rebalancing updates

simusud committed:
files renamed

simusud committed:
test updates

simusud committed:
added bad bebts tests

simusud committed:
loans update

simusud committed:
small changes

simusud committed:
small update

simusud committed:
bad deb test cases

simusud committed:
rebalancing test update

simusud committed:
resolving merge conflicts

simusud committed:
patched bnUSD address

simusud committed:
small change

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
small change

simusud committed:
small change

simusud committed:
resolving merge conflict

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
new rebablancing changes

simusud committed:
new rebablancing changes

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
test updates

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
rebalancing updates

simusud committed:
conflicts fixes

simusud committed:
conflicts fixes

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
small changes

simusud committed:
optimization tests updates

simusud committed:
optimization tests updates

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
retire bad debts tests update

simusud committed:
retireRedeem called through loans tokenFallback update tests

simusud committed:
small change

simusud committed:
solving merge conflicts

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

simusud committed:
rebalancing testnet update

simusud committed:
Merge branch 'rebalancing' into loans-integration-tests

Suyog007 committed:
merged with rebalancing

simusud committed:
retired bnUSD test

Suyog007 committed:
merged with rebalancing

Suyog007 committed:
reduced bnUSD and total bnUSD burnt checked

Suyog007 committed:
Merge branch 'rebalancing' into loans-integration-tests

Suyog007 committed:
rebalancing path location changed

simusud committed:
percent of retired bnUSD from diff users asserted

simusud committed:
mrege conflicts fixes

simusud committed:
tests with multiple addreses

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
Merge branch 'development' of https://github.com/balancednetwork/contracts-private into development

simusud committed:
fixing merge conflicts

simusud committed:
optimised code tests

simusud committed:
Merge branch 'development' into loans-integration-tests

Suyog007 committed:
loans and rebalancing file updated for reverse rebalancing case

Suyog007 committed:
new function created to set bnusd receivable amount in rebalancing

Suyog007 committed:
new interfaces created in governance

Suyog007 committed:
new variable introduced to store bnusd to receive value

simusud committed:
added reverse rebalancing tests

simusud committed:
Merge branch 'rebalancing_reverse_updated' into loans-integration-tests

simusud committed:
changes score update method
updated score update call

simusud committed:
resolving merge conflicts

Suyog007 committed:
merged with rebalancing_updated branch

simusud committed:
rebalancing test update

simusud committed:
solving merge conflicts

simusud committed:
removed prints from base file

simusud committed:
updated MAX_RETIRE_PERCENT on loans
updated rebalancing tests

Suyog007 committed:
reverse rebalancing case updated

simusud committed:
resolving merge conflicts

simusud committed:
removed redundant code in governance
variable name changed in loans
removed data variable from rebaancing
test for forward rebalancing and reverse rabalancing updated

simusud committed:
solving merge conflicts in rebalancing
rebalancing test cases update

simusud committed:
merge conflict fixes
rebalancing test update

simusud committed:
Merge branch 'rebalancing_updated' into loans-integration-tests

simusud committed:
rebalancing tests update

simusud committed:
rebalancing forward and reverse tests changes

simusud committed:
loans update

simusud committed:
updated tests with letest code updates

simusud committed:
changed or to and in tests

simusud committed:
Merge branch 'rebalancing_updated' into loans-integration-tests

simusud committed:
test changes for rebalancing updates

gedanziger committed:
Continuous Rewards

Implementation of continuous rewards. Supports:
- Dex Continuous Rewards interfaces
- Staked LP Contract
- All shared interfaces

Suyog007 committed:
changes in loans for continuous reward

Suyog007 committed:
changes in loans for continuous reward

Suyog007 committed:
check added in collateral_value and to_debt

Suyog007 committed:
rewards functions called from loans contract

Commit:
removal of unused codes and change in new_pos

Suyog007 committed:
updateRewardsData called after burning token

Suyog007 committed:
removal of unused_code and change in new_pos

Commit:
Merge branch 'loans-changes' of https://github.com/balancednetwork/balanced-contracts into loans-changes

Suyog007 committed:
modified the use of type dict

Suyog007 committed:
added interface for rebalancing batch data

Suyog007 committed:
reverted back to previous code

Suyog007 committed:
reverted back to previous code

Suyog007 committed:
changes made in loans for continuous rewards

Suyog007 committed:
Merge branch 'development' into loans-changes

Suyog007 committed:
snapshot file reverted back to original

Suyog007 committed:
liquidate function modified

Suyog007 committed:
interface added in governance.

Suyog007 committed:
new dictdb introduced in positions file.

Suyog007 committed:
type dict modified and minor bugs fixed.

Suyog007 committed:
arraydb out of index bug fixed

dbrehmer committed:
fix route from address

Suyog007 committed:
Merge branch 'continuous-integration' of https://github.com/balancednetwork/balanced-contracts into continuous-integration

Suyog007 committed:
Merge branch 'loans-changes' of https://github.com/balancednetwork/balanced-contracts into loans-changes

Suyog007 committed:
merged with loans-changes branch

Suyog007 committed:
minor change in lowerPrice function

dbrehmer committed:
fix index in conversion loop

dbrehmer committed:
cleared notebook output

simusud committed:
removed  sejong_contract_deployment.ipynb file

Deceivious committed:
Fixes on while loop for route_contract_balances

Deceivious committed:
Fixes on while loop for route_contract_balances

Suyog007 committed:
data migration for loans

gedanziger committed:
[cont-rewards] Make balance available on first claim

Previously @bennyoptions reported only being able to claim rewards via two transactions.

This fixes the issue, so rewards are immediately claimable. To test:

1. Create a fresh deployment, or use a fresh address on a contract with this diff
2. Take out a loan
3. Claim rewards

A transfer will be sent immediately, which wasn't done previously

gedanziger committed:
[cont-rewards] Set continuous rewards day from governance

Adds a function to set the continuous rewards day from the governance SCORE.

Invoke `setContinuousRewardsDay` with param `_day` to set the starting day
for continuous rewards. The SCORE will operate in a non-continuous mode before
that day, then in a continuous mode after that day.

Suyog007 committed:
minor change in to_dict logic

Suyog007 committed:
edge case handled

Suyog007 committed:
Merge branch 'continuous-integration' of https://github.com/balancednetwork/balanced-contracts into continuous-integration

Suyog007 committed:
revert added in readonly functions

Suyog007 committed:
merge with continuous-integration

Suyog007 committed:
getBnusdValue modified

Suyog007 committed:
if asset not active , continuing the loop

Suyog007 committed:
Upgrade process for Continuous Rewards updated

gedanziger committed:
[cont-rewards] Deploy fixes for day1 deployments

Before, the frontend would crash without a few necessary methods in dividends. It required the method for lookup user rewards to return, which it would revert on day 1.

Additionally, it sets the required stakedLp configs in governance when listing new markets, and adds a method to setup the BALN/SICX market.

Deceivious committed:
Added is_dev flag for dev environments
Added set_zero_hour_dev to set zeroth hour of day for dev environments

gedanziger committed:
[draft] Fixes to day context in rewards

Before, the contract was considering the data source platform day, which didn't augument up by one with the introduction of continuous rewards. Instead, this switches the quantity to the actual day.

NOTE: Review with @dbrehmer how we should react if the SCORE isn't called for a day. We may want to preserve the platform day quantity in order to create a prorated emission for each day when computing weight.

Deceivious committed:
fixed zeroth hour for distribution

Deceivious committed:
changed launchBalanced for dev mode

Suyog007 committed:
take snapshot on the day 1

Suyog007 committed:
Merge branch 'continuous-integration' into loans-data-migration

simusud committed:
Merge branch 'loans-integration-tests' into continuous-integration

Suyog007 committed:
merged with continuous integration
 branch

simusud committed:
added base setup file for continuous rewards testing

gedanziger committed:
[fixup] fix edits applied in merge of test branch

simusud committed:
added tear down method on continuous rewards base file

gedanziger committed:
[DEX] DEX Code Size Reduction

Some methods are deprecated/no longer used, and some extra computation was done.
Currently in an upgrade, the operation will fail if the code size is 28kb or higher.
This reduces the code size down to 27.5kb, and saves the users a little bit of cost
on gas fees.

Suyog007 committed:
code modified for sending users info to rewards contract

Suyog007 committed:
getBalanceAndSupply modified

Suyog007 committed:
commented the condition

Suyog007 committed:
standing updated for non continuous rewards

Suyog007 committed:
dev mode removed

Suyog007 committed:
integration test added

Suyog007 committed:
unused import removed

Suyog007 committed:
setup files added

dbrehmer commented on core_contracts/stakedLp/interfaces.py:
```diff
@@ -0,0 +1,60 @@
+class AssetConfig(TypedDict):
+    _id: int
+    asset: Address
+    distPercentage: int
```
How is AssetConfig used? What is an Asset here? Is distPercentage the same as the distPercentage in rewards? If so, how do they stay in sync?

dbrehmer commented on core_contracts/stakedLp/utils/math.py:
```diff
@@ -0,0 +1,40 @@
+        
+    return z
+    
+def convertToExa(_amount:int,_decimals:int)-> int:
```
Needs proper spacing in method definition.

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+
+    @only_admin
+    @external
+    def setGovernance(self, _governance: Address) -> None:
```
May want to require address to be a contract here.

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        Gets the current admin address. This user can call using the
+        `@only_admin` decorator.
+        """
+        return self._governance.get()
```
Should be _admin.

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        # Validate inputs
+        pool_id = self._supportedPools[_id]
+        StakedLp._require(pool_id, f'pool with id:{_id} is not supported')
+        StakedLp._require(_value > 0, f'Cannot stake less than zero ,value to stake {_value}')
```
Comma position in error message - 'zero ,value'

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+
+    @external
+    def unstake(self, _id: int, _value: int) -> None:
+        StakedLp._require(_id in self._supportedPools, f'pool with id:{_id} is not supported')
```
Is this condition sufficient? Couldn't the key be in the DictDB, but the value be False? This is a more readable condition than the one on line 136, but the remove_pool method would need to be changed to delete the DictDB key.

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        StakedLp._require(_id in self._supportedPools, f'pool with id:{_id} is not supported')
+        _user = self.msg.sender
+        StakedLp._require(_value > 0, f'Cannot unstake less than zero'
+                                      f'value to stake {_value}')
```
stake -> unstake

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        super().__init__(db)
+        # Business Logic
+        self._supportedPools = DictDB('supportedPools', db, bool)
+        self._poolStakeDetails = DictDB('poolStakeDetails', db, value_type=int, depth=3)
```
Could we put a comment here saying what the detail keys are for each pool?

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        previous_balance = self._poolStakeDetails[_user][_id][Status.STAKED]
+        previous_total = self._totalStaked[_id]
+
+        StakedLp._require(previous_balance >= _value, f'Cannot unstake,user dont have enough staked balance '
```
Check message grammar and punctuation.

dbrehmer commented on core_contracts/dex/dex.py:
```diff
@@ -442,6 +470,23 @@ def getFeehandler(self) -> Address:
+        :param _address: New contract address to set.
+        Sets new fee handler contract address.
+        """
+        self._stakedLp.set(_address)
```
Maybe we should require a contract address here.

dbrehmer commented on core_contracts/dex/dex.py:
```diff
@@ -442,6 +470,23 @@ def getFeehandler(self) -> Address:
+    @external(readonly=True)
+    def getStakedLp(self) -> Address:
+        """
+        Gets the address of the fee handler contract.
```
'fee handler' -> 'staked LP'

dbrehmer commented on core_contracts/dex/dex.py:
```diff
@@ -551,11 +596,30 @@ def is_locking_pool(self, id: int) -> bool:
+        continuous_rewards_launched = (self._continuous_rewards_day.get() > self._current_day.get())
+        restricted_pool_id = (id < FIRST_NON_BALANCED_POOL or id == USDS_BNUSD_ID or id == IUSDT_BNUSD_ID)
+
+        if  not restricted_pool_id or continuous_rewards_launched:
```
The 'not' here makes this return the opposite of what it currently does. Maybe it should return True on this condition?

gedanziger replied:
I actually fixed this on sejong a week ago and forgot to push, adding now.

dbrehmer commented on core_contracts/dex/dex.py:
```diff
@@ -586,20 +650,24 @@ def fallback(self):
             # TODO: Modify instead of cancel/replace, after debugging scorelib
             node = self._icx_queue._get_node(order_id)
-            order_value += node.get_value1()
+            old_order_value = node.get_value1()
```
Is the change here just for clarity?

dbrehmer replied:
Okay, I see where they are used below to send to rewards.

gedanziger committed:
[cont-rewards] Bugfixes and cleanup for rewards

Before, there was an issue around day rollovers, especially in the presence of multiple days.
Additionally, there was an issue in fliping the condition for enabling transfers.
Returns sum of total weight instead of resetting it.

Deceivious committed:
loans migration method for user

Deceivious committed:
Merge remote-tracking branch 'origin/loans-data-migration' into loans-data-migration

Deceivious committed:
added flag setting

Suyog007 committed:
getBalanceAndSupply modified . No permission to write issue fixed

dbrehmer commented on core_contracts/loans/loans/positions.py:
```diff
@@ -189,26 +192,32 @@ def get_standing(self, _day: int = -1, _readonly: bool = False) -> dict:
-            if bnUSD_debt < self._loans._min_mining_debt.get():
+                bnUSD_debt: int = debt * EXA // price
+                if bnUSD_debt < self._loans._min_mining_debt.get():
+                    standing = Standing.NOT_MINING
```
@Suyog007 - I think we were planning to remove the mining threshold. We can check with Scott, but I think we want give rewards all the way down to liquidation, in which case there is no need for the mining or not_mining states.

Suyog007 replied:
Unless continuous rewards day is activated, we will have all the threshold on the loans and once the continuous rewards day is activated, we will have only mining and liquidation state i.e. all the users that are not on liquidation state will be on mining state.
https://github.com/balancednetwork/balanced-contracts/blob/continuous-integration/core_contracts/loans/loans/positions.py#L195

dbrehmer commented on core_contracts/loans/loans/positions.py:
```diff
@@ -43,8 +43,11 @@ def __getitem__(self, _symbol: str) -> int:
+            day = self.check_snap()
+            self.assets[day][key] = value
+        else:
+            self.assets[self.snaps[-1]][key] = value
```
@Suyog007 - I thought we were going to remove the snaps index and make a dew data structure so all collateral types would be treated the same. Otherwise the structure will always have that unused key and extra level of depth.

Suyog007 replied:
Actually , loans migration is implemented on another branch and it is yet to be merged with contiuous-integration branch.
https://github.com/balancednetwork/balanced-contracts/blob/loans-data-migration/core_contracts/loans/loans/positions.py#L60

Suyog007 committed:
bug fix

Suyog007 committed:
setting string as a key of dictdb

Suyog007 committed:
bug test added

simusud committed:
day changes in const files in loans,dex,rewards
added rebalancing address in data_objects.py
removed dev mode from launchBalanced in governance.py
changes in loans.py and positions.py
added data migration tests

nightowl121 committed:
Remove the previous on_update code

nightowl121 committed:
Switch dividends to staked baln only
- Set a switching day from which the dividends are distributed to staked baln token only
- Dividends after the set day will not use LP tokens for dividends

Deceivious committed:
Merge remote-tracking branch 'origin/continuous-integration' into loans-data-migration

# Conflicts:
#	core_contracts/loans/loans/loans.py

simusud committed:
added rebalancing contract into setup

simusud committed:
removed BALN lp count logic from voting
updated unit test

Suyog007 committed:
functions name renamed

Suyog007 committed:
functions name renamed

simusud committed:
added dividends update procedure in Continuous Rewards Update.ipynb

dbrehmer commented on core_contracts/loans/loans/loans.py:
```diff
@@ -469,12 +484,13 @@ def getDataBatch(self, _name: str, _snapshot_id: int,
             self._current_day.set(day)
             self._positions._take_snapshot()
-            self.check_dead_markets()
+        self.check_dead_markets()
```
Calling check_dead_markets() every time checkForNewDay() is called may be too often. Previously it was only called once per day. is_dead() makes three interscore calls - https://github.com/balancednetwork/balanced-contracts/blob/continuous-integration/core_contracts/loans/loans/assets.py#L140

ajaya-man replied:
Is it okay or should we optimize it?

Suyog007 committed:
fetch total supply in all case

Suyog007 committed:
fetch total supply in all case

Suyog007 committed:
condition modified in is_locking_pool

Suyog007 committed:
test cases added for baln allocation

Suyog007 committed:
condition modified in _update_total_weight

Suyog007 committed:
baln allocation test file added

Suyog007 committed:
Merge branch 'loans-data-migration' into continuous-integration

dbrehmer commented on core_contracts/stakedLp/stakedLp.py:
```diff
@@ -0,0 +1,199 @@
+        """
+        :param _dex: the new DEX address to set.
+        """
+        self._governance.set(_dex)
```
This should be set to governance, not DEX address.

Suyog007 committed:
setters modified

Suyog007 committed:
interface added for dividends setDividendsOnlyToStakedBalnDay function

Suyog007 committed:
getDataCount fn added

simusud committed:
added addPoolOnStakedLp method on governance

ajaya-man commented on core_contracts/loans/loans/positions.py:
```diff
@@ -369,7 +463,8 @@ def _take_snapshot(self) -> None:
         snapshot.snap_time.set(self._loans.now())
         self._loans.Snapshot(self._loans.getDay())
-        self._snapshot_db.start_new_snapshot()
+        if self._loans.getDay() != self._loans._continuous_reward_day.get():
```
@dbrehmer I think we should have < rather than != here. as we want to take snapshot before we switch to continuous rewards day.
